### PR TITLE
Retain imports in declaration emit if they augment an export of the importing file

### LINKED ISSUE #32298: Merged declarations are not written to the parent
**TypeScript Version:**  3.6.0-dev.20190704

**Search Terms:** merge declarations augment grafting

**Code**

parent.ts:

```ts
import { child1 } from './child1';

export class ParentThing implements ParentThing {}

// Add extra methods to this object.
child1(ParentThing.prototype);
// Imagine many more get added as well.
```

child1.ts:

```ts
import { ParentThing } from './parent';

declare module './parent' {
    interface ParentThing {
        add: (a: number, b: number) => number;
    }
}

export function child1(prototype: ParentThing) {
    prototype.add = (a: number, b: number) => a + b;
}
```

test.ts:

```ts
// Use this to test before compilation
import { ParentThing } from './parent';

// Use this to test after compilation
// import { ParentThing } from '../lib/parent';

const parentThing = new ParentThing();

console.log(parentThing.add(1, 1));
```

**Expected behavior:**

When `tsc` compiles the project, I should not lose the definitions. I want to be able to use the generated code as a module and still see the augmented type and be able to follow through to the methods that are defined by all children.

Several children are adding many methods to the parent.

**Actual behavior:**

Before compilation, there is no problem. I can jump to the definition when `test.js` is pointing to `parent.ts`. The declarations are merged and I can immediately access the `add` method in `child1`.

After compilation and when I change the test to load the compiled version, I get an error message: "Property 'add' does not exist on type 'ParentThing'."

**Playground Link:** None.

**Related Issues:** None that I could find.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #32298

Technically this won't fix all possible augmentation-including scenarios, namely this will only capture augmentations of direct dependencies; augmentations provided by transitive dependencies are a whole other bag of problems. But I'm wondering if we actually need to preemptively fix that or not; in such scenarios we'd probably need to generate and add `reference` directives, which could be much more undesirable than the simple preservation of an existing import.

### CODE REVIEW

**Comment by sandersn** [comment]
It would be nice to know how many projects have augmentations coming directly from dependencies vs transitive dependencies. But I can't think of an easy way to estimate that.

**Comment by weswigham** [comment]
When it comes to preserving something already in the declaration file vs injecting something from the aether, I'm also just much more comfortable preserving the existing thing.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -36046,9 +36046,29 @@ namespace ts {
...
 
+            function isImportRequierdByAugmentation(node: ImportDeclaration) {
+                const file = getSourceFileOfNode(node);
+                if (!file.symbol) return false;
+                const importTarget = getExternalModuleFileFromDeclaration(node);
+                if (!importTarget) return false;
+                if (importTarget === file) return false;
+                const exports = getExportsOfModule(file.symbol);
+                for (const s of arrayFrom(exports.values())) {
+                    const merged = getMergedSymbol(s);
```
Comment (sandersn) [inline src/compiler/checker.ts:36061]: this function should return true only if there's a merge, right? So could we check for `if (!s.mergeId) continue` before this line?

**Comment by sandersn** [comment]
Agree!

**Pull request merged**